dashboard/src/components/ParkingRegionsMap.tsx

```diff
@@ -0,0 +1,204 @@
+import * as chroma from 'chroma-js';
+import * as geojson from 'geojson';
+import * as Leaflet from 'leaflet';
+import * as React from 'react';
+import * as ReactLeaflet from 'react-leaflet';
+
+import { MapViewport, Point, Region, RegionProperties } from './types';
+
+import './ParkingRegionsMap.css';
+
+// Add viewport props to the allowed props for Map component, since
+// they are missing from the latest libdefs available.
+// See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/22125
+interface Viewport { center: Point; zoom: number; }
+interface MapProps extends ReactLeaflet.MapProps {
+    onViewportChange?: (viewport: Viewport) => void;
+    onViewportChanged?: (viewport: Viewport) => void;
+    viewport?: Viewport;
+}
+class Map extends ReactLeaflet.Map<MapProps> {}
+
+export interface Props {
+    center: Point;
+    zoom: number;
+    regions?: Region[];
+    onRegionClicked?: (region: Region) => void;
+    onViewportChanged?: (viewport: MapViewport) => void;
+}
+
+export default class ParkingRegionsMap extends React.Component<Props> {
+    private map?: ReactLeaflet.Map;
+
+    render() {
+        const geoJsonElement = this.props.regions ? (
+            <ReactLeaflet.GeoJSON
+                key={getKeyForRegions(this.props.regions)}
+                data={getFeatureCollection(this.props.regions)}
+                style={getStyleForRegion}
+                onEachFeature={this.bindPopupToRegion}
+            />
+        ) : null;
+        const osmAttribution = (
+            '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a>'
+            + ' contributors');
+        return (
+            <Map
+                center={this.props.center}
+                zoom={this.props.zoom}
+                onViewportChanged={this.handleViewportChange}
+                ref={(component: ReactLeaflet.Map) => {
+                        this.map = component; }}
+            >
+                <ReactLeaflet.TileLayer
+                    url="//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
+                    attribution={osmAttribution}
+                />
+                <ReactLeaflet.ScaleControl
+                    metric={true}
+                    imperial={false}
+                    maxWidth={150}
+                />
+                {geoJsonElement}
+            </Map>);
+    }
+
+    private bindPopupToRegion = (
+        region: Region,
+        layer: Leaflet.Layer
+    ): void => {
+        const p = region.properties;
```

Comment:
Style bikeshedding suggestion:
I'm usually not a big fan of using single letter variables in cases where readability suffers a lot. In this case, when `p` is used in many places and not just in a single loop or such, it hurts readability because the reader has no idea wher `p` is pointing to.